### PR TITLE
Compile the limit menu after all other filter panels

### LINKED ISSUE #2932: DCA panels depend on specific order
**Affected Versions**
Contao 4.9

Hello there,

in a recent project we had the issue that `DC_Table` ignored our `filter`and `search` panels for the `limit` panel and back-end pagination. The search by e-mail-address results in three entries, which are correctly listed (not visible in the screenshot). The pagination and the `limit` panel however still take all some hundred entries into account.

The panel configuration in this case was `limit;search;sort`. Because the search panel sets the actual search conditions and values, the limit menu rendered as no search took place. The limit menu then sets the limit for the pagination, hence the pagination is off, too. Defining the panels with `search;sort;limit` fixed this issue.

We should compile the panels in the order necessary to properly work and than reorder them to what was defined by the developer. I'll provide a PR for this as soon as I find the time :-)

Cheers

### PR DESCRIPTION
**Pull request by bezin**
Fixes #2932. Basically if there is a limit menu, I insert a placeholder and skip compiling the menu at first. After all other panels were compiled, especially search and filter, I compile the limit menu and replace the placeholder.

### CODE REVIEW

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 1]
```diff
@@ -1179,8 +1179,16 @@ protected function panel()
 			{
 				$panel = '';
 
+				// The limit menu depends on other panels that may set a filter query, e.g. search and filter.
+				// In order to correctly calculate the total row count, the limit menu must be compiled last.
+				// We isnert a placeholder here and compile the limit menu after all other panels.
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
				// We insert a placeholder here and compile the limit menu after all other panels.
```

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 2]
```diff
@@ -1179,8 +1179,16 @@ protected function panel()
...
+				// In order to correctly calculate the total row count, the limit menu must be compiled last.
+				// We insert a placeholder here and compile the limit menu after all other panels.
+				if ($strSubPanel == 'limit')
+				{
+					// Set placeholder to inject compiled limit menu later
+					$panel = '###limit_menu###';
+				}
 				// Regular panels
-				if ($strSubPanel == 'search' || $strSubPanel == 'limit' || $strSubPanel == 'sort')
+				elseif ($strSubPanel == 'search' || $strSubPanel == 'sort')
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]: Now that nearly all `$strSubPanel` variants are handled individually, we could maybe replace the last two `$panel = $this->{$strSubPanel . 'Menu'}(…);` expressions with real function calls? Took me a while to get what this was doing.

Comment (bezin) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: Yeah, I agree this takes a moment to figure that out. Would you continue with if {} elseif {} in this case?

Comment (m-vo) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: Yes, or a `switch` statement. But let's hear what others think about this first.

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: Doesn't sound like something related to this PR (fixing the limit menu) 😇

Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: What's wrong with `$this->{$strSubPanel . 'Menu'}()`? It is a real function call, isn't it? 😄 

I'm ok with a switch statement as well, though.

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 3]
```diff
@@ -1226,6 +1234,18 @@ protected function panel()
...
+		// Compile limit menu if placeholder is present
+		foreach ($arrPanels as $key => $strPanel)
+		{
+			if (strpos('###limit_menu###', $strPanel) === false)
+			{
+				continue;
+			}
+
+			$limitPanel = $this->limitMenu();
+			$arrPanels[$key] = str_replace('###limit_menu###', $limitPanel, $strPanel);
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
			$arrPanels[$key] = str_replace('###limit_menu###', $this->limitMenu(), $strPanel);
```

No need for a variable here, is there?

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 2 above)
Comment (bezin) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: I refactored this part to a switch statement. Definitely helps with readability.

Comment (m-vo) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: > What's wrong with $this->{$strSubPanel . 'Menu'}()? It is a real function call, isn't it? smile

:smile: I was more referring to things like this where the dynamic part is unnecessary:

```php
elseif ($strSubPanel == 'filter')
{
    $panel = $this->{$strSubPanel . 'Menu'}(++$intFilterPanel);
}
```
&rarr;
```php
elseif ($strSubPanel == 'filter')
{
    $panel = $this->filterMenu(++$intFilterPanel);
}
```

**Review by aschempp — APPROVED** [review]

**Review by Toflar — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bezin.

**Comment by fritzmg** [comment]
It looks like this in the current `4.9` branch for me:



Can anyone confirm?

**Comment by bezin** [comment]
~I cannot. Which view is this? Could you share your DCA config?~

Yes, I can. The parameters passed to `strpos()` are in the wrong order 🙈 Hence, the code does not work if the limit menu is in the same row with another menu. I'll fix that.

**Comment by fritzmg** [comment]
That's the view of a news archive. No customisation.

edit: ok :)

**Pull request merged**